design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+kind: ManualScalerTrait
+metadata:
+  namespace: default
+  name: coolest-app-n3v3y
+spec:
+  replicaCount: 3
+---
+apiVersion: cache.crossplane.io/v1alpha1
+kind: RedisCluster
+metadata:
+  namespace: default
+  name: coolest-app-jnkf9
+spec:
+  classSelector:
+    matchLabels:
+      environment: production
+      region: europe
+  writeConnectionSecretToRef:
+    name: coolest-app-redis
+  engineVersion: "3.2
+```
+
+### Crossplane as an OAM Runtime
+
+This document proposes Crossplane initially support two kinds of component:
+`ContainerizedWorkload`, and Crossplane resource claims. Each containerized
+workload would be rendered to a series of Kubernetes resources (e.g. `Service`,
+`Deployment`, `Ingress`, etc). These resources would be created either in the
+Kubernetes control plane upon which Crossplane is running (similar to Rudr), or
+scheduled to an external Kubernetes cluster of which Crossplane is aware. The
+proposed `SchedulingScope` would determine where the resources were scheduled.
+Each resources that should be created on an external Kubernetes cluster would be
+represented in the Crossplane API server as a `KubernetesApplicationResource` as
+it is today. Note that these `KubernetesApplicationResource` resources would be
+owned (in the controller reference sense) by the `ContainerizedWorkload`; there
+would be no intermediate `KubernetesApplication`. `KubernetesApplication` would
+continue to exist as a distinct, lower level concept in Crossplane, as it is
+useful for scheduling and running arbitrary Kubernetes resources on an external
+cluster without the separation of concerns focused abstraction of an
+`ApplicationConfiguration`.
+
+## Open Questions
+
+The following questions are unresolved by this proposal:
+
+* Should Crossplane resource claims natively support the concept of OAM traits?
+  This proposal implies a distinct Kubernetes controller would reconcile any
+  `ContainerizedWorkload` resources, which would be required to maintain object
+  references to any configured traits. The proposal also implies that resource
+  claims _become_ a kind of extended workload (rather than being produced by a
+  kind of extended workload).
+* Should Crossplane resource claims natively support the concept of OAM scopes?
+  Similar to the above concerns, if an `ApplicationConfiguration` templates and
+  produces resource claims verbatim (as opposed to producing a resource claim
+  modified based on the supplied scopes), the controllers for said claims would
+  need to be aware of scopes in order to consider them when acting on the
+  resource claim.
+* What is the best approach to validate the array of components and traits under
+  the `ApplicationConfiguration` `components` field? One option may be to use a
+  validating webhook that finds the CRD that defines the templated kind and uses
+  its OpenAPI schema to validate it.
+* How could we maintain the existing user experience around easily discovering
```

Comment:
CRD categories will list all resources belong to that category. In fact we just want to list Trait CRDs, so we may need to use label.

```
kubectl get crds -l  oam.dev/part-of=trait
```

Reply:
> CRD categories will list all resources belong to that category.

I see your point that listing the CRDs (the schema) would be closer to today's `kubectl get trait`, given that today traits are schema. It could also be useful to be able to list all instances of traits (even if they were different CR kinds), though.

Reply:
But I think we could still use `WorkloadType` and `Trait` to do register/discovery thing.

Reply:
Traits are meant to be behavioral characteristics on instances of components. It wouldn't make sense to list _instances_ of traits.

Reply:
> It wouldn't make sense to list instances of traits.

Would it make sense if these characteristics were modelled as distinct API objects when instantiated? If I follow correctly you're saying a trait, applied by the app config, might somehow mutate the component instance produced by a schematic. Our thinking was more along the lines that a component with traits (at the app config level) would produce a component instance and several trait instances - there would be relationships (references) between these documents such that the runtime (controller) reconciling the component instance could factor in the trait instances.